packages/mom5/package.py

```diff
@@ -0,0 +1,317 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# Copyright 2023 ACCESS-NRI
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import install, join_path, mkdirp
+
+# https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html
+class Mom5(MakefilePackage):
+    """MOM is a numerical ocean model based on the hydrostatic primitive equations."""
+
+    homepage = "https://www.access-nri.org.au"
+    git = "https://github.com/ACCESS-NRI/mom5.git"
+
+    maintainers = ["harshula"]
+
+    version("spack-build", branch="spack-build")
+
+    # Depend on virtual package "mpi".
+    depends_on("mpi")
+    depends_on("oasis3-mct")
+    depends_on("datetime-fortran")
+    depends_on("netcdf-fortran@4.5.2:")
+    depends_on("netcdf-c@4.7.4:")
+    # TODO: For initial verification we are going to use static pio.
+    #       Eventually we plan to move to shared pio
+    # ~shared requires: https://github.com/spack/spack/pull/34837
+    depends_on("parallelio~pnetcdf~timing~shared")
+    depends_on("libaccessom2")
+    depends_on("hdf5")
+
+    phases = ["edit", "build", "install"]
+
+    def edit(self, spec, prefix):
+
+        srcdir = self.stage.source_path
+        makeinc_path = join_path(srcdir, "bin", "mkmf.template.spack")
+        config = {}
+
+        incs = (spec["oasis3-mct"].headers).cpp_flags + "/psmile.MPI1" + " "
+        for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]:
+            incs += (spec[l].headers).cpp_flags + " "
+
+        # NOTE: The order of the libraries matter during the linking step!
+        # NOTE: datetime-fortran is a dependency of libaccessom2.
+        libs = "-L" + (spec["parallelio"].prefix) + "/lib -lpiof -lpioc "
```

Comment:
```suggestion
        libs = ["-L" + join_path(spec["parallelio"].prefix, "lib"), "-lpiof", "-lpioc"]
```
As above, use a list ...

Reply:
1) That change results in the following compile failure (string vs array):
`  >> 9916    make: *** [Makefile:45: fms_ACCESS-OM.x] Error 1`
2) Regardless, I should have used join_path().

Reply:
I should have said this was untested. Combined with the code below (which suggested using a `join`) I'd have thought it should work, assuming I didn't have a typo, so my apology if that wasted your time.

The thing I really don't think is good practice is appending spaces in the strings, e.g.
```
"-lpiof -lpioc "
```
they're easy to lose or not understand the importance of. 

in another PR you've done something like this:
```
"-lpiof -lpioc" + " "
```
which is preferable.

But ultimately this is why I suggested creating an array and then using `.join` as I think the intent is more explicit: the elements of the array are important tokens, the space is not. Once you put the space in the tokens you're imbuing it with more meaning than it has. Or to put it another way, you don't know if it is meaningful or not. IMO.

Reply:
We've conflated two issues:
1) Constructing a string from a strings.
2) Constructing a string from a list.

For (1), yes join() appears to be a better solution.
For (2), join() appears to traverse the list twice and the suggested approach is too frankensteinish for my liking.

Reply:
Heh. I'd suggest replacing "frankensteinish" with "pythonic". It is a common pattern in python, and so the intent is recognisable.

As I said I really don't mind except I don't think we should have strings with magic spaces at the end that break the functionality if the space is removed. On that we appear to be in agreement, so the rest is up to you. Seriously, do what you makes most sense to you.

Reply:
Here's an unfrankensteinish pythonic approach:
```
        lstr = " ".join(
                    ["".join("-L", join_path(spec["parallelio"].prefix, "lib")),
                    "-lpiof",
                    "-lpioc"]
                   )
        ldeps = ["oasis3-mct", "libaccessom2", "netcdf-c", "netcdf-fortran", "datetime-fortran"]
        libs = " ".join([lstr] + [(spec[d].libs).ld_flags for d in ldeps])
```

Reply:
I like it. Very clean and clear.